File: ios/SpentyAI/SpentyAI/Features/FeatureRequests/FeatureRequestsViewModel.swift
import Foundation
import SwiftUI

@Observable
final class FeatureRequestsViewModel {

    // MARK: - State

    var requests: [FeatureRequest] = []
    var isLoading = false
    var showForm = false
    var errorMessage: String?

    // MARK: - Dependencies

    private let repository: FeatureRequestsRepository

    // MARK: - Init

    init(repository: FeatureRequestsRepository = .shared) {
        self.repository = repository
    }

    // MARK: - Actions

    @MainActor
    func loadRequests() async {
        isLoading = true
        errorMessage = nil

        do {
            requests = try await repository.getAll()
        } catch let error as APIError {
            errorMessage = error.localizedDescription
        } catch {
            errorMessage = "Failed to load feature requests."
        }

        isLoading = false
    }

    @MainActor
    func submitRequest(title: String, description: String, category: FeatureRequestCategory) async {
        errorMessage = nil

        do {
            let newRequest = try await repository.create(
                title: title,
                description: description,
                category: category
            )
            requests.insert(newRequest, at: 0)
            showForm = false
        } catch let error as APIError {
            errorMessage = error.localizedDescription
        } catch {
            errorMessage = "Failed to submit your request."
        }
    }

    @MainActor
    func voteForRequest(id: String) async {
        errorMessage = nil

        do {
            let response = try await repository.vote(id: id)
            if let index = requests.firstIndex(where: { $0.id == id }) {
                let old = requests[index]
                requests[index] = FeatureRequest(
                    id: old.id,
                    title: old.title,
                    description: old.description,
                    category: old.category,
                    status: old.status,
                    votes: response.votes,
                    createdAt: old.createdAt
                )
            }
        } catch let error as APIError {
            errorMessage = error.localizedDescription
        } catch {
            errorMessage = "Failed to register your vote."
        }
    }
}

File: android-native/app/src/main/java/com/spentyai/app/features/featurerequests/FeatureRequestsViewModel.kt
package com.spentyai.app.features.featurerequests

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.spentyai.app.core.models.FeatureRequest
import com.spentyai.app.core.network.ApiResult
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

data class FeatureRequestsUiState(
    val requests: List<FeatureRequest> = emptyList(),
    val isLoading: Boolean = false,
    val showForm: Boolean = false,
    val errorMessage: String? = null
)

class FeatureRequestsViewModel(
    private val repository: FeatureRequestsRepository
) : ViewModel() {

    private val _uiState = MutableStateFlow(FeatureRequestsUiState())
    val uiState: StateFlow<FeatureRequestsUiState> = _uiState.asStateFlow()

    fun loadRequests() {
        viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true, errorMessage = null) }
            when (val result = repository.getAll()) {
                is ApiResult.Success -> {
                    _uiState.update { it.copy(requests = result.data, isLoading = false) }
                }
                is ApiResult.Failure -> {
                    _uiState.update {
                        it.copy(
                            isLoading = false,
                            errorMessage = result.error.message ?: "Failed to load feature requests."
                        )
                    }
                }
            }
        }
    }

    fun submitRequest(title: String, description: String, category: FeatureRequestCategory) {
        viewModelScope.launch {
            _uiState.update { it.copy(errorMessage = null) }
            when (val result = repository.create(title, description, category.value)) {
                is ApiResult.Success -> {
                    _uiState.update {
                        it.copy(
                            requests = listOf(result.data) + it.requests,
                            showForm = false
                        )
                    }
                }
                is ApiResult.Failure -> {
                    _uiState.update {
                        it.copy(errorMessage = result.error.message ?: "Failed to submit your request.")
                    }
                }
            }
        }
    }

    fun voteForRequest(id: String) {
        viewModelScope.launch {
            _uiState.update { it.copy(errorMessage = null) }
            when (val result = repository.vote(id)) {
                is ApiResult.Success -> {
                    _uiState.update { state ->
                        state.copy(
                            requests = state.requests.map { req ->
                                if (req.id == id) result.data else req
                            }
                        )
                    }
                }
                is ApiResult.Failure -> {
                    _uiState.update {
                        it.copy(errorMessage = result.error.message ?: "Failed to register your vote.")
                    }
                }
            }
        }
    }

    fun showForm() {
        _uiState.update { it.copy(showForm = true) }
    }

    fun hideForm() {
        _uiState.update { it.copy(showForm = false) }
    }

    fun dismissError() {
        _uiState.update { it.copy(errorMessage = null) }
    }
}
